### PR TITLE
Rendering build duration 

### LINKED ISSUE #2542: Indicate a build is in progress in Dashboard build page
When I go to a build in GitHub Actions, I get an animated indication that a build is happening, and how long it's been going for:

Screen Shot 2021-11-29 at 9 55 32 AM

When it's done, the yellow animated circle becomes a green check mark or a red 'x'. 

Screen Shot 2021-11-29 at 9 58 45 AM

Let's do something like this in our build dashboard, so you know what's happening when you get there.  Especially when a build step is not producing any output, it can be hard to know if it's still going.

### PR DESCRIPTION
**Pull request by tpmai22**
Fixed #2542 
- Delete animated-indication.css
- Delete duplicate result and duration column in builds.hbs
- Delete datetime-calculator.js and it usage
- Loading animation added
- Adding progress column and showing build progressing animation
  
Co-authored-by: Tengzhen <tengzhenzhao@gmail.com>

To test the feature locally, do the following:

- Change `autodeploymentUrl` in `src\api\status\public\js\build-log\check-build-status.js`

```js
const autodeploymentUrl = (path) => `//dev.telescope.cdot.systems/deploy${path}`;  
```

- Allow all `connectSrc` in `src\api\status\src\server.js`

```js
connectSrc: ["'self'", "*"]
```

To trigger a build manually trigger [redeliver](https://github.com/Seneca-CDOT/telescope/settings/hooks/189238170/deliveries) one of the build which merged to `refs/heads/master`

## Type of Change

- [x] **Bugfix**: Change which fixes an issue
- [ ] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [x] **UI**: Change which improves UI

## Description
Prefer to #2542 

## Checklist

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

**Comment by humphd** [comment]
This is going to need to change based on #2653, where I'm changing a bunch of the code around.  You might want to rebase on that PR's branch and update your code to use what I have there already.  Then when it lands, you can rebase on `master`.

**Comment by tpmai22** [comment]
@Andrewnt219 @humphd would you guys mind to take a look at `datetime-calculator.js` cause I did not see any usage of it even though I did refactor it

File: src/api/status/public/js/build-log/datetime-calculator.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  if (startedAtValue !== null && stoppedAt === null) {
+    buildIcon.innerHTML = 'motion_photos_on';
+    buildIcon.classList.add('fa-spin', 'text-warning', 'fa-12x');
+  } else if (result === 0 && stoppedAt !== null) {
+    buildIcon.innerHTML = 'check_circle';
+    buildIcon.classList.add('text-success');
+  } else if (result !== 0 && stoppedAt !== 'null') {
+    buildIcon.innerHTML = 'cancel';
+    buildIcon.classList.add('text-danger');
+  }
```
Comment (Andrewnt219) [inline src/api/status/public/js/build-log/datetime-calculator.js:36]: These are the needed changes. Move these inside the function that creates the build-header. 

Screenshots or videos for documenting new UI changes are much appreciated (inside PR description).

**Review by Andrewnt219 — CHANGES_REQUESTED** [review]
Don't worry about calculating the duration, it has already been done.

File: src/api/status/public/js/build-log/datetime-calculator.js (hunk 1 above)
Comment (tpmai22) [reply src/api/status/public/js/build-log/datetime-calculator.js:36]: you want me to also delete the `datetime-calculator.js` ?

Comment (Andrewnt219) [reply src/api/status/public/js/build-log/datetime-calculator.js:36]: Yeah, we won't need it.

**Comment by humphd** [comment]
You should rebase this to get the changes I just landed for caching build logs.

File: src/api/status/public/js/build-log/build-header.js [hunk 2]
```diff
@@ -42,11 +42,20 @@ function renderBuildInfo({ isCurrent, githubData, startedDate, stoppedDate, code
...
+    buildResult.innerHTML = 'motion_photos_on';
+    buildResult.classList.add('fa-spin', 'text-warning', 'fa-12x');
+  } else if (code === 0 && stoppedDate !== null) {
+    buildResult.innerHTML = 'check_circle';
+    buildResult.classList.add('text-success');
+  } else if (code !== 0 && stoppedDate !== 'null') {
+    buildResult.innerHTML = 'cancel';
+    buildResult.classList.add('text-danger');
+  }
+
```
Comment (Andrewnt219) [inline src/api/status/public/js/build-log/build-header.js:55]: Icon without text can be ambiguous, so let's keep `buildResult` as text as before, then add another element for icons. 
Also, these should be moved into a function like the others below.

**Review by Andrewnt219 — CHANGES_REQUESTED** [review]
Can you provide screenshots or videos in the PR's description? These are helpful for reviewing and documenting PR changes.

File: src/api/status/public/js/build-log/build-header.js (hunk 2 above)
Comment (tpmai22) [reply src/api/status/public/js/build-log/build-header.js:55]: Right now it kind of hard for me to test it out. I pinged Josue to trigger build manually but they seem not working and not receiving build information. I will try to work on it. Any guidance ?

Comment (Andrewnt219) [reply src/api/status/public/js/build-log/build-header.js:55]: The terminal is showing build-log but the build-header is not showing what is being built?

Comment (tpmai22) [reply src/api/status/public/js/build-log/build-header.js:55]: https://github.com/Seneca-CDOT/telescope/blob/dc273ca972bd1a8684042796a619121da9c9efa7/src/api/status/public/js/build-log/build-header.js#L62 It always went it to this line right away

Comment (humphd) [reply src/api/status/public/js/build-log/build-header.js:55]: What is happening in https://github.com/Seneca-CDOT/telescope/blob/dc273ca972bd1a8684042796a619121da9c9efa7/src/api/status/public/js/build-log/check-for-build.js#L9 ?

Comment (tpmai22) [reply src/api/status/public/js/build-log/build-header.js:55]: I think I pick it up from your PR #2653 . By the way how can you debug it manually without trigger any build ?

Comment (humphd) [reply src/api/status/public/js/build-log/build-header.js:55]: No, I mean when `build` is null, what is being sent to you?  The line I gave you is where it picks the build.

You can't debug it without doing it for real, which is what I was saying on Slack yesterday about needing to find a way to do it..

Comment (tpmai22) [reply src/api/status/public/js/build-log/build-header.js:55]: I pinged you on Slack

**Comment by tpmai22** [comment]
@humphd Hitting the `/deploy/status` gives me 404. I think the autodeploye url for developement is wrong. Here is my shared workspace, can you confirm it?
https://tpmai22-telescope-5ekqsq7bamt.ws-us27.gitpod.io/

**Comment by tpmai22** [comment]
404 on both production and dev.
image

**Comment by humphd** [comment]
Yeah, that's not going to work in gitpod I guess?  I'm not sure if the autodeployment server gets run in gitpod, it would have to be started manually, it's not in a container.

**Comment by tpmai22** [comment]
Any guidance to run it. I went into `tools\autodeployment` then `pnpm start` but it say something about `secrect`

**Comment by Andrewnt219** [comment]
So far, there's no way to locally start the autodeployment. In the previous PRs (see test description of #2582 for example), the workaround is pointing to the dev URL instead of the local.

So the only way to test it is to get everything ready and trigger the build manually (by Joshua or David). In your case, I believe the logic is doable to implement without checking the result all the time.

You said the build didn't work, can you elaborate more on what was not working? 
- Did the build log (terminal) show any text about the current build?
- Did the build header show the latest commit info, duration, and status?

**Comment by tpmai22** [comment]
I did talk with David and I believe that was because of gitpod set up I will try to run it locally and see

**Comment by tpmai22** [comment]
I learn how to test it locally in #2582 and here is the result 

https://user-images.githubusercontent.com/43376325/150909276-c644301e-f12c-4b94-a7a8-993c823e7d3f.mp4

**Comment by Andrewnt219** [comment]
Awesome! Just one UI fix, what I meant was having the text next to the icon (to indicate what the icon means), not in a separate column.

I'd love to test this out locally as well, but the redeliver link in PR description is broken.

File: src/api/status/public/js/build-log/build-header.js [hunk 3]
```diff
@@ -29,6 +30,31 @@ function renderSha(compare, after) {
...
+  if (startedDate != null && stoppedDate == null) {
+    buildLoadingAnimation.innerHTML = 'motion_photos_on';
+    buildLoadingAnimation.classList.add('fa-spin', 'fa-5x', 'text-success', 'fa');
+  } else if (code === 0 && stoppedDate != null) {
+    buildLoadingAnimation.innerHTML = 'check_circle';
+    buildLoadingAnimation.classList.add('text-success');
+  } else if (code !== 0 && stoppedDate != null) {
+    buildLoadingAnimation.innerHTML = 'cancel';
+    buildLoadingAnimation.classList.add('text-danger');
```
Comment (Andrewnt219) [inline src/api/status/public/js/build-log/build-header.js:44]: Question. If the build is in-progress, does it have `code`? If not, we only need `code` for this switch.

**Comment by tpmai22** [comment]
@Andrewnt219 you can go to `setting\webhook` and `resend` one of the build that merged to `master`

**Comment by Andrewnt219** [comment]
The settings is probably only available to OSD700 students. I can review them and someone else can test it.

**Comment by tpmai22** [comment]
@Andrewnt219 After talking on Slack with Andrew, we decide to delete the text part and keep only the loading animation in the `build-header`
image

File: src/api/status/public/js/build-log/build-header.js [hunk 4]
```diff
@@ -29,6 +29,24 @@ function renderSha(compare, after) {
   buildGitSHA.innerText = after.substring(0, 7);
 }
 
+function renderLoadingAnimation(code) {
+  buildLoadingAnimation.class = 'material-icons';
```
Comment (Andrewnt219) [inline src/api/status/public/js/build-log/build-header.js:33]: This should be [`className`](https://developer.mozilla.org/en-US/docs/Web/API/Element/className).

File: src/api/status/public/js/build-log/build-header.js [hunk 5]
```diff
@@ -29,6 +29,24 @@ function renderSha(compare, after) {
...
+
+  switch (code) {
+    case 0:
```
Comment (Andrewnt219) [inline src/api/status/public/js/build-log/build-header.js:36]: Nit: Document these if you can. 0, 1, and default seem like magic numbers.

**Review by Andrewnt219 — CHANGES_REQUESTED** [review]
Only a small typo left and it's good to go.

File: src/api/status/public/js/build-log/build-header.js (hunk 5 above)
Comment (tpmai22) [reply src/api/status/public/js/build-log/build-header.js:41]: @Andrewnt219 I made some comments on top of it.

**Review by Andrewnt219 — APPROVED** [review]
Awesome, thanks for picking up this PR.

Need someone from OSD700 to follow the PR's test plan and test this locally.

File: src/api/status/public/js/build-log/build-header.js [hunk 6]
```diff
@@ -29,6 +29,29 @@ function renderSha(compare, after) {
   buildGitSHA.innerText = after.substring(0, 7);
 }
 
+function renderLoadingAnimation(code) {
+  buildLoadingAnimation.className = 'material-icons';
+
+  /*
+  Build success return code === 0
+  Build failure  return code === 1
```
Comment (humphd) [inline src/api/status/public/js/build-log/build-header.js:37]: Any non-zero result code is a failure, it won't necessarily be 1.  I think you should do:

```js
if(code === 0) { /* success */ }

if(!code) { /* no code present (i.e., it's not zero, so it's undefined/null), unknown (i.e., in progress) */ }

/* anything else is a failure */
```

**Review by humphd — CHANGES_REQUESTED** [review]

**Comment by tpmai22** [comment]
This will also close #2562

**Review by Andrewnt219 — APPROVED** [review]

**Review by humphd — APPROVED** [review]

**Pull request merged**